### ISSUE TITLE
[Windows] vitest exits silently without watching or printing coverage when using canvas

### ISSUE THREAD

**Issue by kleinfreund** [vitest-dev/vitest#2261]
### Describe the bug

(It seems the issue described here applies to Windows only. All failing cases listed below pass on Ubuntu via WSL2 in Windows but fail in Windows PowerShell.)

For *some* runs of vitest,

* it doesn't watch. It does print the watch output (i.e. "Waiting for file changes...", etc.).
* it doesn't print a coverage report. It does print the coverage output (i.e. "% Coverage report from c8").

### Reproduction

1. Clone https://github.com/kleinfreund/vue-accessible-color-picker/tree/a650e7fd451e1ea23eeb2f95ce4d29ec8183b3c2 (sorry, can't reproduce this in a more minimal case but the repo should be small enough), run `npm install`.
2. Set `test.singleThread` to `true` in vite.config.js.
3. Run the following commands:
   - `npx vitest` <br>❌ doesn't watch
   - `npx vitest .\src\utilities\` <br>❌ doesn't watch
   - `npx vitest --coverage` <br>❌ doesn't watch, doesn't print coverage report
   - `npx vitest --coverage .\src\utilities\` <br>❌ doesn't watch, doesn't print coverage report
   - `npx vitest run --coverage` <br>❌ doesn't print coverage report
   - `npx vitest run --coverage .\src\utilities\` <br>❌ doesn't print coverage report
   - `npx vitest .\src\utilities\color-conversions\` <br>✅ watches
   - `npx vitest --coverage .\src\utilities\color-conversions\` <br>✅ watches, prints coverage report
   - `npx vitest run --coverage .\src\utilities\color-conversions\` <br>✅ prints coverage report

#### More context

* Running tests using wildcards in path segments doesn't seem to work whatsoever. All the following runs produce the "No test files found, exiting with code 1" error. Path separator doesn't seem to matter; filter, include, exclude, and watch exclude values seem to all be correct and correctly normalized. I suspect this is some sort of lack of path resolving against the current working directory in Windows environments.
  * `npx vitest run --coverage .\src` <br>✅
  * `npx vitest run --coverage .\src\**\*.test.js` <br>❌ prints "No test files found, exiting with code 1"
  * `npx vitest run --coverage .\src\utilities\*.test.js` <br>❌ prints "No test files found, exiting with code 1"
  * `npx vitest run --coverage .\src\utilities\**\*.test.js` <br>❌ prints "No test files found, exiting with code 1"
* This is not related to Vue. I've ported the package to a separate web component and the exact same issue occurs in that project as well.

### System Info

```shell
System:
    OS: Windows 10 10.0.22621
    CPU: (16) x64 Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
    Memory: 53.80 GB / 63.92 GB
  Binaries:
    Node: 20.5.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.19 - C:\Program Files\nodejs\yarn.CMD
    npm: 9.8.1 - ~\dev\vue-accessible-color-picker\node_modules\.bin\npm.CMD
  npmPackages:
    vite: ^4.4.8 => 4.4.8
    vitest: ^0.34.1 => 0.34.1
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by sheremet-va** [comment]
(Vitest won't clear the terminal in the next version)

**Comment by kleinfreund** [comment]
I've narrowed this down to likely be an issue on Windows specifically. I cannot reproduce any of the failing cases provided above on Ubuntu (via WSL2). I've updated the original report accordingly.

**Comment by sheremet-va** [comment]
Is it still an issue?

**Comment by kleinfreund** [comment]
@sheremet-va I’m afraid so. I’ve checked this again a few days ago and updated the original report with (newer) affected versions.

**Comment by sheremet-va** [comment]
> Running tests using wildcards in path segments doesn't seem to work whatsoever. All the following runs produce the "No test files found, exiting with code 1" error. Path separator doesn't seem to matter; filter, include, exclude, and watch exclude values seem to all be correct and correctly normalized

This depends on the shell you are using. Linux automatically wraps this and passed down an array of files. Vitest only does regexp with these files.

**Comment by mendrik** [comment]
I have same problem but maybe this stack traces helps
```
> vitest run --silent --coverage

Error: Cannot find module './lib/coverage-map'
Require stack:
- C:\desk\*****\***-*****\node_modules\istanbul-lib-coverage\index.js
- C:\desk\*****\***-*****\node_modules\c8\lib\report.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1039:15)
    at Module._load (node:internal/modules/cjs/loader:885:27)
    at Module.require (node:internal/modules/cjs/loader:1105:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at Object.<anonymous> (C:\desk\*****\***-*****\node_modules\istanbul-lib-coverage\index.js:16:25)
    at Module._compile (node:internal/modules/cjs/loader:1218:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1272:10)
    at Module.load (node:internal/modules/cjs/loader:1081:32)
    at Module._load (node:internal/modules/cjs/loader:922:12)
    at Module.require (node:internal/modules/cjs/loader:1105:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    'C:\\desk\\*****\\***-*****\\node_modules\\istanbul-lib-coverage\\index.js',
    'C:\\desk\\*****\\***-*****\\node_modules\\c8\\lib\\report.js'
  ]
}
```

**Comment by kleinfreund** [comment]
@mendrik Not sure this is caused by the same underlying issue. I'm not getting any errors when running the commands above. For me, it prints the regular test output and then:

```
 % Coverage report from c8
```

After which it simply exits.

**Comment by kleinfreund** [comment]
After updating several dependencies in my project, I can no longer reproduce this issue. All commands reported as failing above now behave as expected. The following dependencies related to this issue were updated:

- @vitest/coverage-c8 ^0.28.3 -> ^0.29.1
- c8 ^7.12.0 -> ^7.13.0
- vite ^4.0.4 -> ^4.1.4
- vitest ^0.28.3 -> ^0.29.1

**Comment by kleinfreund** [comment]
I'm reopening this because this issue is only addressed as long as I don't enable `--single-thread`.

As part of https://github.com/vitest-dev/vitest/releases/tag/v0.29.0, #740 was fixed by offering the alternative to set `--single-thread` instead of `--no-threads`:

> If you were using --no-threads before, you might consider enabling --single-thread instead [...]

Now, while keeping `--no-threads` in my project, the issues reported here cannot be reproduced. However, switching to `--single-thread` re-introduces them.

**Comment by kleinfreund** [comment]
Trying to find where this breaks in the vitest code, stepping through code in the vendor-node bundle.

I'm running some of the commands from my original post above in both a Windows and an Ubuntu environment.

I noticed that in `createThreadsPool`, the call `await pool.run(data, ...)` doesn't always resolve before the test run is over (i.e. the process exits).

https://github.com/vitest-dev/vitest/blob/a08e59930bc0667024b26ef94a3e0b5821f97bf0/packages/vitest/src/node/pools/threads.ts#L93

- Ubuntu, without `--coverage`: does always resolve
- Ubuntu, with `--coverage`: does **never** resolve (process exits before, `port.close()` in finally branch is never called)
- Windows, without `--coverage`: does **not always** resolve (process exits before, `port.close()` in finally branch is never called)
- Windows, with `--coverage`: does **never** resolve (process exits before, `port.close()` in finally branch is never called)

When `pool.run` doesn't resolve, the process exits meaning the rest of the function won't be executed (neither will the `post.close` call in the finally branch). That explains why the coverage report isn't always printed.

This all sounds like a race condition somewhere. Maybe something in `Tinypool` or the interactions with it goes wrong?

**Comment by kleinfreund** [comment]
I think I figured out why this happens.

One of the files in my project ([src/utilities/parse-props-color.js](https://github.com/kleinfreund/vue-accessible-color-picker/blob/main/src/utilities/parse-props-color.js)) uses canvas (`document.createElement('canvas').getContext('2d')`). Removing the tests that import this file fixes the issue. I have the [canvas](https://www.npmjs.com/package/canvas) package installed in the project, too.

It's surprising that the tests don't crash or throw an error of a sort. How come it just silently exits?

**Comment by BretHudson** [comment]
<img width="460" alt="image" src="https://github.com/vitest-dev/vitest/assets/1000405/d2baa084-8f24-4f64-930a-9f64b86cdcbf">

I am seeing similar behavior when using `jsdom` and `canvas`! Vitest runs a subset of my tests and then crashes, outputting `Exit status 3221225477`. I am on Windows, and am just calling code I wrote in the test, no third-party libraries/frameworks/packages.

Commenting out the following in my imported code resolves the crash:

```js
const pixelCanvas =
	typeof OffscreenCanvas !== 'undefined'
		? new OffscreenCanvas(1, 1)
		: document.createElement('canvas');
const _pixelCtx = pixelCanvas.getContext('2d');
if (!_pixelCtx) {
	throw Error('pixelCtx failed to create');
}
const pixelCtx = _pixelCtx;
```

Ping if you think having access to the code would help & I can gladly send it over :)

**Comment by AriPerkkio** [comment]
The `canvas` package is unstable when run in `node:worker_threads`. And also the old [`--no-threads` options seems to have fixed](https://github.com/vitest-dev/vitest/issues/2261#issuecomment-1465236461) this so I'll close this issue with following:

Fixed by https://github.com/vitest-dev/vitest/pull/3925 and https://github.com/vitest-dev/vitest/pull/4172.

Vitest 1.0.0-beta has now `--pool=forks` which uses multiple `node:child_process` in parallel. The previous `--no-threads` was using a single `node:child_process`.

If the code you are testing is incompatible with `node:worker_threads`, switch to `--pool=forks`. If you run into same error with that pool, feel free to open new issue with minimal reproduction.

**Comment by BretHudson** [comment]
> Vitest 1.0.0 has now `--pool=forks` which uses multiple `node:child_process` in parallel. The previous `--no-threads` was using a single `node:child_process`.

`--pool=forks` worked perfectly for me, thank you!
